File: pocs/integrations/src/graph/state-meta.ts
import { quotedIdentifier as qi } from "@duckdb/node-api";
import type { QueryableStore } from "../staging/types.js";

export const META_TABLE = "_state/meta";

// Version of the content-hash algorithm for newly written hashes. Bump when the
// hash input or serialization changes; old-hash rows reclassify as updates once
// (PATCH-idempotent) and converge.
export const HASH_VERSION = 2;

export type MetaScope = {
  scope: "entity" | "link";
  connectorId: string;
  sinkId: string;
};

export type SinkMeta = {
  hashVersion: number | null;
  configHash: string | null;
  graphIdentity: string | null;
  webId: string | null;
  namespace: string | null;
};

export async function ensureMetaTable(db: QueryableStore): Promise<void> {
  await db.exec(`CREATE TABLE IF NOT EXISTS ${qi(META_TABLE)} (
    scope VARCHAR,
    connector_id VARCHAR,
    sink_id VARCHAR,
    hash_version INTEGER,
    config_hash VARCHAR,
    graph_identity VARCHAR,
    web_id VARCHAR,
    namespace VARCHAR,
    updated_at TIMESTAMP
  )`);
}

export async function readMeta(db: QueryableStore, s: MetaScope): Promise<SinkMeta | null> {
  await ensureMetaTable(db);
  const { rows } = await db.query(
    `SELECT hash_version, config_hash, graph_identity, web_id, namespace
     FROM ${qi(META_TABLE)}
     WHERE scope = ${esc(s.scope)} AND connector_id = ${esc(s.connectorId)} AND sink_id = ${esc(s.sinkId)}`,
  );
  if (rows.length === 0) return null;
  const r = rows[0];
  return {
    hashVersion: r.hash_version == null ? null : Number(r.hash_version),
    configHash: str(r.config_hash),
    graphIdentity: str(r.graph_identity),
    webId: str(r.web_id),
    namespace: str(r.namespace),
  };
}

export async function writeMeta(db: QueryableStore, s: MetaScope, meta: SinkMeta): Promise<void> {
  await ensureMetaTable(db);
  await db.exec(
    `DELETE FROM ${qi(META_TABLE)}
     WHERE scope = ${esc(s.scope)} AND connector_id = ${esc(s.connectorId)} AND sink_id = ${esc(s.sinkId)}`,
  );
  await db.exec(
    `INSERT INTO ${qi(META_TABLE)} VALUES ($1, $2, $3, $4::INTEGER, $5, $6, $7, $8, now())`,
    [
      s.scope,
      s.connectorId,
      s.sinkId,
      meta.hashVersion == null ? null : String(meta.hashVersion),
      meta.configHash,
      meta.graphIdentity,
      meta.webId,
      meta.namespace,
    ],
  );
}

function esc(v: string): string {
  return `'${v.replace(/'/g, "''")}'`;
}

function str(v: unknown): string | null {
  return v == null ? null : String(v);
}

File: pocs/integrations_rs/src/graph/state_meta.rs
//! Per-sink state metadata in `_state/meta` (same table as the TS/Elixir
//! engines): hash algorithm version, config hash, and the target fingerprint.
//! hash_version applies to newly written hashes; old-hash rows reclassify as
//! updates once and converge.

use error_stack::Report;
use serde_json::Value;

use crate::error::StoreError;
use crate::store::{lit, qi, Store};

const TABLE: &str = "_state/meta";
pub const HASH_VERSION: i64 = 2;

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Meta {
    pub hash_version: Option<i64>,
    pub config_hash: Option<String>,
    pub graph_identity: Option<String>,
    pub web_id: Option<String>,
    pub namespace: Option<String>,
}

pub async fn ensure_table(store: &Store) -> Result<(), Report<StoreError>> {
    store
        .exec(&format!(
            "CREATE TABLE IF NOT EXISTS {} (scope VARCHAR, connector_id VARCHAR, sink_id VARCHAR, hash_version INTEGER, config_hash VARCHAR, graph_identity VARCHAR, web_id VARCHAR, namespace VARCHAR, updated_at TIMESTAMP)",
            qi(TABLE)
        ))
        .await
}

pub async fn read(
    store: &Store,
    scope: &str,
    connector_id: &str,
    sink_id: &str,
) -> Result<Option<Meta>, Report<StoreError>> {
    ensure_table(store).await?;

    let result = store
        .query(&format!(
            "SELECT hash_version, config_hash, graph_identity, web_id, namespace FROM {} WHERE scope = {} AND connector_id = {} AND sink_id = {}",
            qi(TABLE),
            lit(scope),
            lit(connector_id),
            lit(sink_id)
        ))
        .await?;

    Ok(result.rows.first().map(|row| Meta {
        hash_version: row.first().and_then(Value::as_i64),
        config_hash: row.get(1).and_then(Value::as_str).map(str::to_owned),
        graph_identity: row.get(2).and_then(Value::as_str).map(str::to_owned),
        web_id: row.get(3).and_then(Value::as_str).map(str::to_owned),
        namespace: row.get(4).and_then(Value::as_str).map(str::to_owned),
    }))
}

pub async fn write(
    store: &Store,
    scope: &str,
    connector_id: &str,
    sink_id: &str,
    meta: &Meta,
) -> Result<(), Report<StoreError>> {
    ensure_table(store).await?;

    store
        .exec(&format!(
            "DELETE FROM {} WHERE scope = {} AND connector_id = {} AND sink_id = {}",
            qi(TABLE),
            lit(scope),
            lit(connector_id),
            lit(sink_id)
        ))
        .await?;

    let opt = |value: &Option<String>| value.clone().map(Value::from).unwrap_or(Value::Null);
    store
        .exec_params(
            &format!(
                "INSERT INTO {} VALUES ($1, $2, $3, $4, $5, $6, $7, $8, now())",
                qi(TABLE)
            ),
            vec![
                Value::from(scope),
                Value::from(connector_id),
                Value::from(sink_id),
                meta.hash_version.map(Value::from).unwrap_or(Value::Null),
                opt(&meta.config_hash),
                opt(&meta.graph_identity),
                opt(&meta.web_id),
                opt(&meta.namespace),
            ],
        )
        .await
}
